util/test/demos/d3d11/d3d11_test.cpp

```diff
@@ -114,65 +114,50 @@ void D3D11GraphicsTest::Prepare(int argc, char **argv)
       ID3D11DevicePtr tempDev;
 
       D3D_FEATURE_LEVEL features[] = {D3D_FEATURE_LEVEL_11_0};
-      hr = dyn_D3D11CreateDevice(adapter, D3D_DRIVER_TYPE_HARDWARE, NULL, 0, features, 1,
-                                 D3D11_SDK_VERSION, &tempDev, NULL, NULL);
+      hr = CreateDevice(adapters, NULL, features, 0);
 
       if(SUCCEEDED(hr))
       {
-        tempDev->CheckFeatureSupport(D3D11_FEATURE_D3D11_OPTIONS, &opts, sizeof(opts));
+        dev->CheckFeatureSupport(D3D11_FEATURE_D3D11_OPTIONS, &opts, sizeof(opts));
 
         ID3D11Device1Ptr tempDev1;
 
-        tempDev1 = tempDev;
+        tempDev1 = dev;
         memset(&opts1, 0, sizeof(opts1));
         if(tempDev1)
           tempDev1->CheckFeatureSupport(D3D11_FEATURE_D3D11_OPTIONS1, &opts1, sizeof(opts1));
       }
+
+      // This device was only used  to get feature support. Set it back to NULL
+      dev = NULL;
     }
   }
 }
 
-bool D3D11GraphicsTest::Init()
+bool D3D11GraphicsTest::Init(IDXGIAdapter **ppAdapter)
```

Comment:
Could this be `IDXGIAdapterPtr` with no indirection instead? It doesn't seem useful to have the ability to deliberately override a NULL adapter given what we've found, so we could only apply this override if the parameter is non-NULL.

Reply:
NULL can also mean using a warp adapter, so I wanted to be able to clarify between overriding to warp and not overriding. If you think that's not worth worrying about (i.e., if D3D12 selects warp, will D3D11 always select warp too?) then I can switch to `IDXGIAdapterPtr`

Reply:
Ah right. For this case I'd favour simplicity over robustness. I think either WARP is being forced on the command line, in which case both APIs should pick WARP, or there are no D3D capable adapters at all (really?!) in which case both APIs should fall back to WARP.

If we really come across a situation where we need to force WARP and this is broken we can evaluate then, but for this demo program I don't think it's worth worrying about.